ledger/ledger-api-tests/infrastructure/src/main/scala/com/daml/ledger/api/testtool/infrastructure/participant/SingleParticipantTestContext.scala

```diff
@@ -0,0 +1,846 @@
+import com.daml.platform.participant.util.HexOffset
+import com.daml.platform.testing.StreamConsumer
+import com.google.protobuf.ByteString
+import io.grpc.health.v1.health.{HealthCheckRequest, HealthCheckResponse}
+import io.grpc.stub.StreamObserver
+import scalaz.Tag
+import scalaz.syntax.tag._
+
+import scala.concurrent.duration.DurationInt
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.Failure
+import scala.util.control.NonFatal
+
+private[testtool] object SingleParticipantTestContext {
+
+  private[this] def filter(templateIds: Seq[Identifier]): Filters =
+    new Filters(
+      if (templateIds.isEmpty) None
+      else Some(new InclusiveFilters(templateIds))
+    )
+
+  private def transactionFilter(
+      parties: Seq[String],
+      templateIds: Seq[Identifier],
+  ): Some[TransactionFilter] =
+    Some(new TransactionFilter(Map(parties.map(_ -> filter(templateIds)): _*)))
+
+}
+
+/** Exposes services running on some participant server in a test case.
+  *
+  * Each time a test case is run it receives a fresh instance of [[ParticipantTestContext]]
+  * (one for every used participant server).
+  */
+final class SingleParticipantTestContext private[participant] (
+    val ledgerId: String,
+    val endpointId: String,
+    val applicationId: String,
+    val identifierSuffix: String,
+    override val referenceOffset: LedgerOffset,
+    protected[participant] val services: LedgerServices,
+    partyAllocationConfig: PartyAllocationConfiguration,
+    override val ledgerEndpoint: Endpoint,
+    override val features: Features,
+)(protected[participant] implicit val ec: ExecutionContext)
+    extends UserManagementTestContext
+    with ParticipantTestContext {
+  private val logger = ContextualizedLogger.get(getClass)
+
+  import SingleParticipantTestContext._
+
+  private[this] val identifierPrefix =
+    s"$applicationId-$endpointId-$identifierSuffix"
+
+  private[this] def nextIdGenerator(name: String, lowerCase: Boolean = false): () => String = {
+    val f = Identification.indexSuffix(s"$identifierPrefix-$name")
+    if (lowerCase)
+      () => f().toLowerCase
+    else
+      f
+  }
+
+  private[this] val nextPartyHintId: () => String = nextIdGenerator("party")
+  private[this] val nextCommandId: () => String = nextIdGenerator("command")
+  private[this] val nextSubmissionId: () => String = nextIdGenerator("submission")
+  private[this] val workflowId: String = s"$applicationId-$identifierSuffix"
+  override val nextKeyId: () => String = nextIdGenerator("key")
+  override val nextUserId: () => String = nextIdGenerator("user", lowerCase = true)
+
+  override lazy val delayMechanism: DelayMechanism = if (features.staticTime) {
+    new StaticTimeDelayMechanism(this)
+  } else
+    new TimeDelayMechanism()
+
+  override def toString: String = s"participant $endpointId"
+
+  /** Gets the absolute offset of the ledger end at a point in time. Use [[end]] if you need
```

Comment:
Scaladoc is duplicated in multiple places. I think it should reside in the interface if it's general, or in the implementing class if it's related to the implementation.